### ISSUE TITLE
Assertion fails on `armv7` when built with `NEON`

### ISSUE THREAD

**Issue by jpalus** [BLAKE3-team/BLAKE3#350]
`blake3` (`c` implementation) built for `linux/amrv7` with  `cmake` and`NEON` support enabled fails on following assert (when used in `mold`):
```
ld.mold: /home/users/builder/rpm/BUILD/BLAKE3-1.5.0/c/blake3.c:343: compress_subtree_to_parent_node: Assertion `num_cvs <= MAX_SIMD_DEGREE_OR_2' failed.
```

It appears to be caused by compilation definition `BLAKE3_USE_NEON` being defined only for `blake3_dispatch.c`:
https://github.com/BLAKE3-team/BLAKE3/blob/5aa53f07f7188a569cadfc5daf1522972d9a9630/c/CMakeLists.txt#L134-L134

But not on `blake3.c` which uses `blake3_impl.h` and causing `MAX_SIMD_DEGREE` mismatch:
https://github.com/BLAKE3-team/BLAKE3/blob/5aa53f07f7188a569cadfc5daf1522972d9a9630/c/blake3_impl.h#L51-L70

**Comment by oconnor663** [comment]
@jpalus thanks for the report. Your diagnosis seems plausible to me! @BurningEnlightenment can you suggest a `CMakeLists.txt` fix? Maybe we should also add this case to CI.

**Comment by BurningEnlightenment** [comment]
The fix should be as simple as
```
diff --git a/c/CMakeLists.txt b/c/CMakeLists.txt
--- c/CMakeLists.txt
+++ c/CMakeLists.txt
@@ -130,9 +130,9 @@
 
   target_sources(blake3 PRIVATE
     blake3_neon.c
   )
-  set_source_files_properties(blake3_dispatch.c PROPERTIES COMPILE_DEFINITIONS BLAKE3_USE_NEON=1)
+  set_source_files_properties(blake3.c blake3_dispatch.c PROPERTIES COMPILE_DEFINITIONS BLAKE3_USE_NEON=1)
 
   if (DEFINED BLAKE3_CFLAGS_NEON)
     set_source_files_properties(blake3_neon.c PROPERTIES COMPILE_FLAGS "${BLAKE3_CFLAGS_NEON}")
   endif()

```
> Maybe we should also add this case to CI.

I don't really fancy getting into the cross compiling business, but I can probably submit a PR later this week.

**Comment by jpalus** [comment]
Since README.md recommends building all sources with `-DBLAKE3_USE_NEON=1`:
https://github.com/BLAKE3-team/BLAKE3/blob/5aa53f07f7188a569cadfc5daf1522972d9a9630/c/README.md?plain=1#L278-L281
wouldn't it be better to add it to all sources when building with `cmake`?
```diff
diff --git a/c/CMakeLists.txt b/c/CMakeLists.txt
index c73b9c7..116ea58 100644
--- a/c/CMakeLists.txt
+++ b/c/CMakeLists.txt
@@ -131,7 +131,7 @@ elseif(CMAKE_SYSTEM_PROCESSOR IN_LIST BLAKE3_ARMv8_NAMES
   target_sources(blake3 PRIVATE
     blake3_neon.c
   )
-  set_source_files_properties(blake3_dispatch.c PROPERTIES COMPILE_DEFINITIONS BLAKE3_USE_NEON=1)
+  add_definitions(-DBLAKE3_USE_NEON=1)
 
   if (DEFINED BLAKE3_CFLAGS_NEON)
     set_source_files_properties(blake3_neon.c PROPERTIES COMPILE_FLAGS "${BLAKE3_CFLAGS_NEON}")
```

**Comment by BurningEnlightenment** [comment]
That's something @oconnor663 needs to decide, I've kept the build description as close to their original form as possible when I modernized them. 

Note that `add_definitions()` has been superseeded by `add_compile_definitions()`, however, that is only available from v3.12 onwards, so it should be spelled `target_compile_definitions(blake3 PRIVATE BLAKE3_USE_NEON=1)` instead. (Modifying directory level properties with `add*_definitions()` is discouraged in modern CMake, anyways.)

**Comment by oconnor663** [comment]
Yes, preprocessor defines like `BLAKE3_USE_NEON` and `BLAKE3_NO_AVX512` should be set the same way for all files in the build. But this is in contrast to CPU feature flags like `-mavx2` that we use for building intrinsics, which must only be set for the one matching code file. Maybe we should call both of these things out explicitly in `c/README.md`?
